### PR TITLE
feat(streaming): watermark on append only table

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Allow define a watermark on append only table created with source.

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by st1page — APPROVED** [review]
LGTM! we can add some e2e tests to test if we can filter the outdated data. (also fine in next pr)

File: src/frontend/planner_test/tests/testdata/watermark.yaml [hunk 1]
```diff
@@ -13,3 +13,13 @@
         └─StreamRowIdGen { row_id_index: 1 }
           └─StreamWatermarkFilter { watermark_descs: [idx: 0, expr: (v1 - '00:00:01':Interval)] }
             └─StreamSource { source: "t", columns: ["v1", "_row_id"] }
+- name: watermark on append only table
+  sql: |
+    explain create table t (v1 timestamp with time zone, watermark for v1 as v1 - INTERVAL '1' SECOND) with (connector = 'kafka', kafka.topic = 'kafka_3_partition_topic', kafka.brokers = '127.0.0.1:1234', kafka.scan.startup.mode='earliest', appendonly=true) ROW FORMAT JSON;
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/frontend/planner_test/tests/testdata/watermark.yaml:18]: I think we can keep both behaviors undocumented for test. I'll add a new planner test.

**Review by xx01cyx — COMMENTED** [review]
Generally LGTM. BTW where is the "append only" part of this PR?  👀

File: src/frontend/src/handler/create_table.rs [hunk 2]
```diff
@@ -438,12 +465,21 @@ fn gen_table_plan_inner(
...
     }
 
+    if !append_only && !watermark_descs.is_empty() {
+        return Err(ErrorCode::NotSupported(
+            "Defining watermarks on table requires the table to be append only.".to_owned(),
+            "Set the option `appendonly=true`".to_owned(),
+        )
+        .into());
+    }
+
```
Comment (yuhao-su) [inline src/frontend/src/handler/create_table.rs:475]: @xx01cyx append only part

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 3]
```diff
@@ -356,9 +358,11 @@ impl ToBatch for LogicalSource {
 impl ToStream for LogicalSource {
     fn to_stream(&self, _ctx: &mut ToStreamContext) -> Result<PlanRef> {
         let mut plan: PlanRef = StreamSource::new(self.clone()).into();
-        if let Some(catalog) = self.source_catalog() && !catalog.watermark_descs.is_empty(){
+        if let Some(catalog) = self.source_catalog() && !catalog.watermark_descs.is_empty() && !self.core.for_table{
             plan = StreamWatermarkFilter::new(plan, catalog.watermark_descs.clone()).into();
         }
+
+        assert!(!(self.core.gen_row_id && self.core.for_table));
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:365]: May add some comments here.

**Pull request merged**